src/docfx/build/Build.cs

```diff
@@ -162,33 +162,65 @@ void ValidateBookmarks()
                 }
 
                 var hasErrors = context.Report(file.ToString(), errors);
-                if (model != null && !hasErrors)
+                if (hasErrors || model == null)
                 {
-                    if (model is string str)
-                        context.WriteText(str, file.OutputPath);
-                    else
-                        context.WriteJson(model, file.OutputPath);
-
-                    return (false, dependencies);
+                    manifestBuilder.MarkError(file);
```

Comment:
how about extracting manifest related logic into an inline method for readability?

Reply:
yeah, the refactor hasn't been completed yet, especially around the legacy part, it'll be in another PR.